### PR TITLE
SQL Assessment Database Selector

### PR DESCRIPTION
**Pull request by cssuh**
- Add the ability to select databases to run assessments
- Display assessment results for selected databases

db-selector
Screen Shot 2021-07-07 at 1 18 52 PM
Screen Shot 2021-07-07 at 1 18 58 PM

### CODE REVIEW

**Comment by brian-harris** [comment]
should the checkbox(select all) and the column headers in step 2 be left aligned?

File: extensions/sql-migration/src/wizard/sourceConfigurationPage.ts [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+	'overflow': 'hidden',
+};
+export class SourceConfigurationPage extends MigrationWizardPage {
+	private _view!: azdata.ModelView;
+	private _databaseSelectorTable!: azdata.DeclarativeTableComponent;
+	private _dbNames!: string[];
+	private _dbCount!: azdata.TextComponent;
+
+	constructor(wizard: azdata.window.Wizard, migrationStateModel: MigrationStateModel) {
+		super(wizard, azdata.window.createWizardPage(constants.SOURCE_CONFIGURATION, 'MigrationModePage'), migrationStateModel);
```
Comment (aasimkhan30) [inline extensions/sql-migration/src/wizard/sourceConfigurationPage.ts:35]: Change the title of page to databases for migration

File: extensions/sql-migration/src/wizard/sourceConfigurationPage.ts [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+	}
+
+	protected async registerContent(view: azdata.ModelView): Promise<void> {
+		this._view = view;
+
+		const flex = view.modelBuilder.flexContainer().withLayout({
```
Comment (aasimkhan30) [inline extensions/sql-migration/src/wizard/sourceConfigurationPage.ts:41]: Add filter resource search bar and fix the spacing 

image

File: extensions/sql-migration/src/wizard/sourceConfigurationPage.ts [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+	}
+
+	protected async handleStateChange(e: StateChangeEvent): Promise<void> {
+	}
+
+
+	public async createRootContainer(view: azdata.ModelView): Promise<azdata.FlexContainer> {
+		const queryProvider = azdata.dataprotocol.getProvider<azdata.QueryProvider>((await this.migrationStateModel.getSourceConnectionProfile()).providerId, azdata.DataProviderType.QueryProvider);
+		const query =
+			`IF OBJECT_ID('tempdb.dbo.#space') IS NOT NULL
```
Comment (aasimkhan30) [inline extensions/sql-migration/src/wizard/sourceConfigurationPage.ts:70]: I would recommend moving trying the queries on older sql servers to make sure we are back compat.

File: extensions/sql-migration/src/wizard/sourceConfigurationPage.ts [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+											ELSE s.compressed_backup_size
+								END / 1048576.0 AS DECIMAL(18,2))
+					, RowNum = ROW_NUMBER() OVER (PARTITION BY s.database_name, s.[type] ORDER BY s.backup_finish_date DESC)
+				FROM msdb.dbo.backupset s
+				WHERE s.[type] IN ('D', 'L')
+			) f
+			WHERE f.RowNum = 1
+			GROUP BY f.database_name
+		) bu ON d.name = bu.database_name
+		ORDER BY t.total_size DESC`;
```
Comment (aasimkhan30) [inline extensions/sql-migration/src/wizard/sourceConfigurationPage.ts:141]: This seems like a big query. I wonder if we should move this STS. Maybe there is something already in STS that we can use. 

image

I am just concerned about breaking the extension due to some edge cases in the query.

**Review by aasimkhan30 — CHANGES_REQUESTED** [review]

File: extensions/sql-migration/src/wizard/sourceConfigurationPage.ts (hunk 1 above)
Comment (cssuh) [reply extensions/sql-migration/src/wizard/sourceConfigurationPage.ts:35]: So removing the (s) from database(s)?

Comment (Charles-Gagnon) [reply extensions/sql-migration/src/wizard/sourceConfigurationPage.ts:35]: Generally it's preferred to just use plurals whenever there can be 1+ of something - (s) is difficult to translate into other languages being one reason.

**Comment by aasimkhan30** [comment]
@cssuh  
`
		let currentConnection = await azdata.connection.getCurrentConnection();
		const metaDataService = azdata.dataprotocol.getProvider<azdata.MetadataProvider>(currentConnection.providerId, azdata.DataProviderType.MetadataProvider);
		const database = await metaDataService.getDatabases(ownerUri);
`

Use this method instead of the query to get database details.

File: extensions/sql-migration/src/wizard/databaseSelectorPage.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	protected async handleStateChange(e: StateChangeEvent): Promise<void> {
+	}
+
+
+	public async createRootContainer(view: azdata.ModelView): Promise<azdata.FlexContainer> {
+		const providerId = (await this.migrationStateModel.getSourceConnectionProfile()).providerId;
+		const metaDataService = azdata.dataprotocol.getProvider<azdata.MetadataProvider>(providerId, azdata.DataProviderType.MetadataProvider);
+		const ownerUri = await azdata.connection.getUriForConnection(this.migrationStateModel.sourceConnectionId);
+		const results = <azdata.DatabaseInfo[]>await metaDataService.getDatabases(ownerUri);
+		// const results = await queryProvider.runQueryAndReturn(await (azdata.connection.getUriForConnection(this.migrationStateModel.sourceConnectionId)), query);
```
Comment (brian-harris) [inline extensions/sql-migration/src/wizard/databaseSelectorPage.ts:72]: can the commented code be removed

File: extensions/sql-migration/src/constants/strings.ts [hunk 6]
```diff
@@ -434,7 +434,7 @@ export const SERVER = localize('sql.migration.server', "Server");
 export const USERNAME = localize('sql.migration.username', "Username");
 export const SIZE = localize('sql.migration.size', "Size (MB)");
 export const LAST_BACKUP = localize('sql.migration.last.backup', "Last backup");
-export const DATABASE_FOR_MIGRATION = localize('sql.migration.database.migration', "Database(s) for Migration");
+export const DATABASE_FOR_MIGRATION = localize('sql.migration.database.migration', "Databases for Migration");
 export const DATABASE_MIGRATE_TEXT = localize('sql.migrate.text', "Select database(s) that you want to migrate to Azure SQL");
```
Comment (brian-harris) [inline extensions/sql-migration/src/constants/strings.ts:438]: for consistency, I think the label "Databases for Migration" should be sentenced cased as 'Databases for migration'

Comment (cssuh) [reply extensions/sql-migration/src/constants/strings.ts:438]: Since it is a title, shouldn't the "Migration" part be capitalized?

Comment (brian-harris) [reply extensions/sql-migration/src/constants/strings.ts:438]: sorry it's ok to leave for now, after looking at other parts of ADS, it seems inconsistent how to apply the capitalization.

**Comment by cssuh** [comment]
@aasimkhan30 @brian-harris I have updated the screenshot in the PR comment.

File: extensions/sql-migration/src/wizard/databaseSelectorPage.ts [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+			width: '100%'
+		}).component();
+		flex.addItem(await this.createRootContainer(view), { flex: '1 1 auto' });
+
+		await view.initializeModel(flex);
+	}
+
+	public async onPageEnter(): Promise<void> {
+		this.wizard.registerNavigationValidator((pageChangeInfo) => {
+			return true;
```
Comment (aasimkhan30) [inline extensions/sql-migration/src/wizard/databaseSelectorPage.ts:64]: Shouldn't you check at least 1 database is selected?

File: extensions/sql-migration/src/wizard/databaseSelectorPage.ts [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+			CSSStyles: {
+				'width': '200px',
+				'margin': '10px 8px 0px 0px'
+			}
+		}).component();
+
+		return searchContainer;
+	}
+
+	@debounce(500)
```
Comment (aasimkhan30) [inline extensions/sql-migration/src/wizard/databaseSelectorPage.ts:97]: What is the purpose of this line?

**Review by aasimkhan30 — APPROVED** [review]
Looks good overall. Please add the validation that I have mentioned before merging it in.

**Comment by aasimkhan30** [comment]
@cssuh. You will also have to update the STS version in this PR before merging it in.

File: extensions/sql-migration/src/wizard/databaseSelectorPage.ts (hunk 8 above)
Comment (brian-harris) [reply extensions/sql-migration/src/wizard/databaseSelectorPage.ts:97]: it throttles the execution of the "decorated" method to only fire after no activity for 500 ms... preventing the search from updating if the user types quickly

File: extensions/mssql/src/contracts.ts [hunk 9]
```diff
@@ -1021,6 +1022,7 @@ export interface SqlAssessmentResult extends azdata.ResultStatus {
 
 export interface SqlMigrationAssessmentParams {
 	ownerUri: string;
+	databases?: string[];
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/contracts.ts:1025]: This is required on the STS side to actually do anything, any reason you have it optional here?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 10]
```diff
@@ -177,7 +176,7 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 
 	public async getDatabaseAssessments(): Promise<ServerAssessement> {
 		const ownerUri = await azdata.connection.getUriForConnection(this.sourceConnectionId);
-
+		// stress test backend & dialog component
 		const assessmentResults = await this.migrationService.getAssessments(
```
Comment (brian-harris) [inline extensions/sql-migration/src/models/stateMachine.ts:180]: is the comment still relevant?

File: extensions/sql-migration/src/models/stateMachine.ts [hunk 11]
```diff
@@ -162,28 +162,37 @@ export class MigrationStateModel implements Model, vscode.Disposable {
 		this._currentState = newState;
 		this._stateChangeEventEmitter.fire({ oldState, newState: this.currentState });
 	}
-
-	public async getServerAssessments(): Promise<ServerAssessement> {
+	public async getDatabases(): Promise<string[]> {
```
Comment (brian-harris) [inline extensions/sql-migration/src/models/stateMachine.ts:165]: can excludeDbs be defined once and shared

**Review by brian-harris — APPROVED** [review]

**Pull request merged**